packages/effective-module-tree/index.js

```diff
@@ -0,0 +1,109 @@
+const fs = require( 'fs' ).promises;
+const path = require( 'path' );
+const globby = require( 'globby' );
+const debug = require( 'debug' )( 'effective-module-tree' );
+const treeify = require( 'object-treeify' );
+
+// Generator that yields all places were node will look for modules
+// For a path like : a/node_modules/b/node_modules/c
+// It will generate the candidates:
+//		- a/node_modules/b/node_modules/c/node_modules
+// 		- a/node_modules/b/node_modules
+//		- a/node_modules
+//		- node_modules
+const candidates = function*( packagePath ) {
+	const parts = packagePath.split( path.sep );
+	for ( let i = parts.length; i >= 0; i-- ) {
+		// This avoid generating .../node_modules/node_modules/...
+		if ( parts[ i - 1 ] === 'node_modules' ) continue;
+		yield path.join( ...parts.slice( 0, i ), 'node_modules' );
+	}
+};
+
+// Simplify tree. Once the full tree has been discovered and linked, we need
+// to simplify it (i.e. remove unused properties) to generate a format compatible
+// with treeify so we can render it (https://www.npmjs.com/package/object-treeify)
+const simplify = ( packageDef, parents ) => {
+	// This means the dependency couldn't be found. Maybe a missing `npm install` ?
+	if ( ! packageDef ) return;
+
+	const name = `${ packageDef.name }@${ packageDef.version }`;
+
+	// Don't get caught in an infinite loop caused by circular dependencies
+	if ( parents.includes( name ) ) return { [ name ]: '[Circular]' };
+
+	if ( ! packageDef.linkedDeps ) return { [ name ]: {} };
+
+	// Keep simplifying our deps recursively
+	const parentDeps = [ ...parents, name ];
+	return {
+		[ name ]: packageDef.linkedDeps.reduce(
+			( accumulated, dep ) => ( { ...simplify( dep, parentDeps ), ...accumulated } ),
+			{}
+		),
+	};
+};
+
+/**
+ * How this works:
+ *
+ * - Find all package.json files
+ * - Read them and store it in a map indexed by path
+ * - Iterate the map, and for each element, iterate the dependencies
+ * - For each dependency, try to find it looking at parent directories
+ * - Once everything is linked, produce a simplified map suited for treeify
+ * - Print it
+ */
+const effectiveTree = async () => {
+	// Find all package.json files
+	const packageJsonPaths = await globby( '**/package.json' );
+
+	// Read all package.json files
+	const packagesByPath = new Map();
+	await Promise.all(
+		packageJsonPaths.map( async packageJsonPath => {
```

Comment:
My `wp-calypso` tree has 5300 `package.json` files. Reading them all in parallel doesn't sound very efficient 😉 

The `**/package.json` glob will also find some directories that are not managed by NPM:
- files like `client/signup/package.json` that don't define a package, but are used to set different main files for browser and Node.js (`main` vs `browser` field) or to set `sideEffects: false` for a module.
- the `test/e2e` directory has its own independent tree that requires a separate `npm install` and has its own `package-lock.json`. We did that to prevent `puppetteer` Chrome download -- but that's no longer a concern, I guess.

Reply:
About perf, it reads them in less than 1 second in my laptop.

Totally agree about `package.json` that are not "real" packages. How do you propose I fix it? Adding a CLI flag to specify directories to exclude?

Reply:
I have more data about the performance:

Running in parallel (`await Promise.all(...)`), it reads and parses all package.json files in 600.662 ms. Running in series (`for ( const p of paths ) { await fs.readFile(...)}`) runs in 828.6924 ms. 

(average of 5 runs, measured using console.time() api)